pkg/operators/wasm/fields.go

```diff
@@ -0,0 +1,192 @@
+// - stack[2]: Kind
+// Return value:
+// - Uint8 value
+func (i *wasmOperatorInstance) fieldAccessorGet(ctx context.Context, m wapi.Module, stack []uint64) {
+	acc, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.FieldAccessor)
+	if !ok {
+		i.logger.Warnf("field handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	data, ok := i.getHandle(wapi.DecodeU32(stack[1])).(datasource.Data)
+	if !ok {
+		i.logger.Warnf("data handle %d not found", stack[1])
+		stack[0] = 0
+		return
+	}
+
+	switch api.Kind(stack[2]) {
+	case api.Kind_Int8:
+		stack[0] = uint64(acc.Int8(data))
+	case api.Kind_Int16:
+		stack[0] = uint64(acc.Int16(data))
+	case api.Kind_Int32:
+		stack[0] = uint64(acc.Int32(data))
+	case api.Kind_Int64:
+		stack[0] = uint64(acc.Int64(data))
+	case api.Kind_Uint8:
+		stack[0] = uint64(acc.Uint8(data))
+	case api.Kind_Uint16:
+		stack[0] = uint64(acc.Uint16(data))
+	case api.Kind_Uint32:
+		stack[0] = uint64(acc.Uint32(data))
+	case api.Kind_Uint64:
+		stack[0] = uint64(acc.Uint64(data))
+	case api.Kind_Float32:
+		stack[0] = uint64(acc.Float32(data))
+	case api.Kind_Float64:
+		stack[0] = uint64(acc.Float64(data))
+	case api.Kind_String:
+		// These are a bit special as they don't fit in the return value, so we have to
+		// allocate an array in the guest memory and return a pointer to it.
+		str := []byte(acc.String(data))
+		res, err := i.guestMalloc.Call(ctx, uint64(len(str)))
+		if err != nil {
+			i.logger.Warnf("malloc failed: %v", err)
+			stack[0] = 0
+			return
+		}
+
+		if !m.Memory().Write(uint32(res[0]), str) {
+			i.logger.Warnf("out of memory write")
+			stack[0] = 0
+			return
+		}
+
+		stack[0] = uint64(len(str))<<32 | uint64(res[0])
+	default:
+		i.logger.Warnf("unknown field kind: %d", stack[2])
+		stack[0] = 0
+	}
+}
+
+// fieldAccessorSet sets the field's value
+// Params:
+// - stack[0]: Field handle
+// - stack[1]: Data handle
+// - stack[2]: Kind
+// - stack[3]: Value to store
+func (i *wasmOperatorInstance) fieldAccessorSet(ctx context.Context, m wapi.Module, stack []uint64) {
+	acc, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.FieldAccessor)
+	if !ok {
+		i.logger.Warnf("field handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	data, ok := i.getHandle(wapi.DecodeU32(stack[1])).(datasource.Data)
+	if !ok {
+		i.logger.Warnf("data handle %d not found", stack[1])
+		stack[0] = 0
+		return
+	}
+
+	switch api.Kind(stack[2]) {
+	case api.Kind_Int8:
+		acc.PutInt8(data, int8(stack[3]))
+		stack[0] = uint64(acc.Int8(data))
+	case api.Kind_Int16:
+		acc.PutInt16(data, int16(stack[3]))
+	case api.Kind_Int32:
+		acc.PutInt32(data, int32(stack[3]))
+	case api.Kind_Int64:
+		acc.PutInt64(data, int64(stack[3]))
+	case api.Kind_Uint8:
+		acc.PutUint8(data, uint8(stack[3]))
+	case api.Kind_Uint16:
+		acc.PutUint16(data, uint16(stack[3]))
+	case api.Kind_Uint32:
+		acc.PutUint32(data, uint32(stack[3]))
+	case api.Kind_Uint64:
+		acc.PutUint64(data, uint64(stack[3]))
```

Comment:
Need to make sure that these won't crash 😄

Reply:
#2823 added logic to allocate the memory and also some tests that hopefully will catch crashes on those ones.